Mathlib/LinearAlgebra/Matrix/Swap.lean

```diff
@@ -0,0 +1,166 @@
+
+@[simp]
+lemma map_swap {S : Type*} [Semiring S] (f : R →+* S) (i j : n) :
+    (swap R i j).map f = swap S i j := by
+  simp [swap]
+
+variable [Fintype n]
+
+lemma swap_mulVec_single (i j : n) (r : R) :
+    swap R i j *ᵥ Pi.single i r = Pi.single j r := by
+  simp only [swap, Equiv.toPEquiv_toMatrix_mulVec_single, Equiv.symm_swap]
+  rw [Equiv.swap_apply_left]
+
+/-- Variant of `Matrix.swap_mulVec_single` with `i` and `j` switched in `Pi.single`. -/
+lemma swap_mulVec_single' (i j : n) (r : R) :
+    swap R i j *ᵥ Pi.single j r = Pi.single i r := by
+  rw [swap_eq_swap, swap_mulVec_single]
+
+lemma swap_mulVec_single_of_ne {i j k : n} (hik : k ≠ i) (hjk : k ≠ j) (r : R) :
+    swap R i j *ᵥ Pi.single k r = Pi.single k r := by
+  simp only [swap, Equiv.toPEquiv_toMatrix_mulVec_single, Equiv.symm_swap,
+    Equiv.swap_apply_of_ne_of_ne hik hjk]
+
+lemma swap_mulVec_apply (i j : n) (a : n → R) :
+    (swap R i j *ᵥ a) i = a j := by
+  simp [swap, Equiv.toPEquiv_toMatrix_mulVec_apply]
+
+/-- Multiplying with `swap R i j` on the left swaps the `i`-th row with the `j`-th row. -/
+@[simp]
+lemma swap_mul_apply (a i j : n) (g : Matrix n n R) :
+    (swap R i j * g) i a = g j a := by
+  simp [swap, PEquiv.toPEquiv_mul_matrix]
+
+/-- Multiplying with `swap R i j` on the left swaps the `j`-th row with the `i`-th row. -/
+@[simp]
+lemma swap_mul_apply' (a i j : n) (g : Matrix n n R) :
+    (swap R i j * g) j a = g i a := by
+  rw [swap_eq_swap, swap_mul_apply]
+
+lemma swap_mul_of_ne {a b i j : n} (hai : a ≠ i) (haj : a ≠ j) (g : Matrix n n R) :
+    (swap R i j * g) a b = g a b := by
+  simp [swap, PEquiv.toPEquiv_mul_matrix, Equiv.swap_apply_of_ne_of_ne hai haj]
+
+/-- Multiplying with `swap R i j` on the right swaps the `i`-th column with the `j`-th column. -/
+@[simp]
+lemma mul_swap_apply (a i j : n) (g : Matrix n n R) :
+    (g * swap R i j) a i = g a j := by
+  simp [swap, PEquiv.mul_toPEquiv_toMatrix]
+
+/-- Multiplying with `swap R i j` on the right swaps the `j`-th column with the `i`-th column. -/
+@[simp]
+lemma mul_swap_apply' (a i j : n) (g : Matrix n n R) :
+    (g * swap R i j) a j = g a i := by
+  rw [swap_eq_swap, mul_swap_apply]
+
+lemma mul_swap_of_ne {a b : n} {i j : n} (hbi : b ≠ i) (hbj : b ≠ j) (g : Matrix n n R) :
+    (g * swap R i j) a b = g a b := by
+  simp [swap, PEquiv.mul_toPEquiv_toMatrix, Equiv.swap_apply_of_ne_of_ne hbi hbj]
+
+/-- Swap matrices are self inverse. -/
+lemma swap_mul_self (i j : n) : swap R i j * swap R i j = 1 := by
+  simp only [swap]
+  rw [← Equiv.swap_inv, Equiv.Perm.inv_def]
+  apply Equiv.toPEquiv_toMatrix_mul_symm_toPEquiv_toMatrix
+
+end
+
+namespace GeneralLinearGroup
+
+variable (R : Type*) {n : Type*} [CommRing R] [DecidableEq n] [Fintype n]
+
+/-- `Matrix.swap` as an element of `GL n R`. -/
+@[simps]
+def swap (i j : n) : GL n R where
+  val := Matrix.swap R i j
+  inv := Matrix.swap R i j
+  val_inv := swap_mul_self i j
+  inv_val := swap_mul_self i j
```

Comment:
If you insist, I can define this in the `PEquiv` setting as well, this has nicer def-eqs though since here `val = inv` is def-eq.